src/glang/codegen/builtin_callables.py

```diff
@@ -698,6 +700,198 @@ def assert_can_write_to(self) -> None:
+    def __init__(
+        self,
+        specialization: list[SpecializationItem],
+        arguments: list[TypedExpression],
+        meta: Meta,
+    ) -> None:
+        assert len(specialization) == 0
+        assert len(arguments) == 0
+        StaticTypedExpression.__init__(
+            self,
+            VoidType(),
+            Type.Kind.VALUE,
+            meta,
+        )
+
+    def __repr__(self) -> str:
+        return "TrapExpression()"
+
+    def generate_ir(self, ctx: IRContext) -> IROutput:
+        ir = IROutput()
+        dbg = self.add_di_location(ctx, ir)
+        ir.lines.append(f"call void @llvm.trap(), {dbg}")
+        return ir
+
+    @property
+    def ir_ref_without_type_annotation(self) -> str:
+        assert False
+
+    def assert_can_read_from(self) -> None:
+        assert False
+
+    def assert_can_write_to(self) -> None:
+        assert False
+
+
+class UnreachableExpression(BuiltinCallable):
+    def __init__(
+        self,
+        specialization: list[SpecializationItem],
+        arguments: list[TypedExpression],
+        meta: Meta,
+    ) -> None:
+        assert len(specialization) == 0
+        assert len(arguments) == 0
+        StaticTypedExpression.__init__(
+            self,
+            VoidType(),
+            Type.Kind.VALUE,
+            meta,
+        )
+
+    def __repr__(self) -> str:
+        return "UnreachableExpression()"
+
+    def generate_ir(self, ctx: IRContext) -> IROutput:
+        ir = IROutput()
+        dbg = self.add_di_location(ctx, ir)
+        ir.lines.append(f"unreachable, {dbg}")
+        return ir
+
+    @override
+    def is_return_guaranteed(self) -> bool:
+        return True
+
+    @property
+    def ir_ref_without_type_annotation(self) -> str:
+        assert False
+
+    def assert_can_read_from(self) -> None:
+        assert False
+
+    def assert_can_write_to(self) -> None:
+        assert False
+
+
+class VolatileWriteExpression(BuiltinCallable):
+    def __init__(
+        self,
+        specialization: list[SpecializationItem],
+        arguments: list[TypedExpression],
+        meta: Meta,
+    ) -> None:
+        assert len(specialization) == 0
+        assert len(arguments) == 2
+
+        (self._addr, self._to_write) = arguments
+        assert self._addr.result_type.storage_kind.is_reference()
+        assert not self._to_write.result_type.storage_kind.is_reference()
+
+        assert (
+            self._addr.result_type.convert_to_value_type() == self._to_write.result_type
+        )
+
+        super().__init__(
+            VoidType(),
+            Type.Kind.VALUE,
+            meta,
+        )
+
+    def __repr__(self) -> str:
+        return f"VolatileWriteExpression({self._to_write} to {self._addr})"
+
+    def generate_ir(self, ctx: IRContext) -> IROutput:
+        # Remove any indirect references
+        conv_to_write, extra_exprs_to_write = do_implicit_conversion(
+            self._to_write, self._to_write.result_type
+        )
```

Comment:
Did we not have a helper for that?

Reply:
I'm not sure. I have a fuzzy memory of one... but also I can't find it... maybe this is already the helper?

Reply:
What about `dereference_double_indirection()`?

Reply:
That seems like a misuse of the utility since we only have (at most) a single layer of indirection here.

This would manifest if we tried something like `__builtin_volatile_write(&addr, 6)` you don't want `dereference_double_indirection` to try and deference `6`. I think they'd produce identical IR for anything else tho.

Reply:
Yeah, you're right